torch/distributed/fsdp/sharded_grad_scaler.py

```diff
@@ -281,27 +281,35 @@ def unscale_(self, optimizer: torch.optim.Optimizer) -> None:
         # Synchronize the detected inf across the ranks
         optimizer_state = self._per_optimizer_states[id(optimizer)]
         future_handles = []
-
-        for v in optimizer_state["found_inf_per_device"].values():
-            if v.device.type == "cpu":
-                v_on_cuda = v.cuda()
+        found_inf_on_cpus = []
+        found_inf_on_cudas = []
+
+        for found_inf in optimizer_state["found_inf_per_device"].values():
+            if found_inf.device.type == "cpu":
+                found_inf_on_cpus.append(found_inf)
+                found_inf_on_cuda = found_inf.cuda()
+                found_inf_on_cudas.append(found_inf_on_cuda)
                 future_handles.append(
                     dist.all_reduce(
-                        v_on_cuda, async_op=True, group=self.process_group
+                        found_inf_on_cuda, async_op=True, group=self.process_group
                     ).get_future()
                 )
-                v.copy_(v_on_cuda.cpu())
             else:
                 future_handles.append(
                     dist.all_reduce(
-                        v, async_op=True, group=self.process_group
+                        found_inf, async_op=True, group=self.process_group
                     ).get_future()
                 )
 
         # Make sure that the calls are done before moving out.
         if future_handles:
             torch.futures.wait_all(future_handles)
```

Comment:
This is not from this PR, but I am not sure I understand why we are using futures and not just using `Work` objects. I imagine that we might be unnecessarily incurring CPU overheads by using futures for these one-element all-reduces.

Reply:
is there a wait_all equivalent for [work1, work2, ....] ?

Reply:
Discussed offline:
```
for work in works:
    work.wait()
```
(In theory, only waiting on the `work` object of the last all-reduce suffices since they are all using the same NCCL PG, but for readability, it could be better to wait on all of them.)